### ISSUE TITLE
Catch HTTPS Connect

### ISSUE THREAD

**Issue by amlwwalker** [elazarl/goproxy#107]
So I can catch the HTTPS Connect, and I started looking at doing this with abourget's fork, however I cant work out how to kill the request and send a response if its over 443.
I think I need to make a custom `ConnectAction` 
So:

```
        proxy.OnRequest().HandleConnectFunc(func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
            log.Println("connecting: ", ctx.Req.URL.Host)
        return goproxy.OkConnect, host
    })
```

No I want to stop the connection to the server and just respond to the client with some HTML, i.e a `&http.Response` but that doesnt seem to be built into the `HandleConnectFunc` as an option.

Is this possible, or am I missing a point?

All I want to do is catch the connect, if the Host contains HTTPS/443, then stop the connect and respond with an HTTP response.

EDIT: Hmmm, I can edit the request but I think once HTTPS happens, then thats that????

If I change the request to be over 80, rather than 443, it still seems to successfully manage the TLS handshake. I expected it to get passed to my proxy.OnResponse function that handles HTTP responses...

```
        proxy.OnRequest().HandleConnectFunc(func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
            log.Println("connecting: ", ctx.Req.URL.Host)
            if strings.Contains(ctx.Req.URL.Host, ":443") {
                log.Println("yup 443")
                ctx.Req.URL.Host = strings.Replace(ctx.Req.URL.Host, "443", "80", -1)
                log.Println("HOST: ", ctx.Req.URL)
            }

        return goproxy.OkConnect, host
    })
```

EDIT AGAIN:

Is there such thing as a CONNECT response from the server? I am forcing all properties of ctx that have :443 over to :80 which results in:
![image](https://cloud.githubusercontent.com/assets/2920180/7818360/24b6dee2-03d3-11e5-874c-75c7b497ae56.png)

Thats almost good ;)
I can stop the HTTPS connection occuring but I cant force a new HTML page to the client. Is that even possible?

**Comment by amlwwalker** [comment]
I feel sooo close:

In the browser I go to `gmail.com`
My code looks like:

```
	proxy.OnRequest().HandleConnectFunc(func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
			log.Println("connecting: ", ctx.Req.URL.Host)
			if strings.Contains(ctx.Req.URL.Host, ":443") {
				ctx.Req.URL.Host = "www.nytimes.com"
				host = ctx.Req.URL.Host
				ctx.Req.RequestURI = host
				ctx.Req.URL.Scheme = "http"
			}
		return goproxy.OkConnect, host
	})
```
In the browser it hangs indefinately as I have changed all of the references off 443 and all references to the host to `www.nytimes.com`.

In the brower I see it hanging indefinately, but it doe NOT put the HTTPS in front of the url I went to, which makes me feel that the HSTS has not yet successfully kicked in:

![image](https://cloud.githubusercontent.com/assets/2920180/7819229/1efcc4e4-03d8-11e5-9661-58cbbe5781ad.png)

This feels like I'm **almost** there. 
All I need to understand (I think) is what the goproxy.OkConnect, is actually doing. I can see in the code it is returning some TLS configuration. 
What I would like to do is just return an HTTP response - as in the browser no HTTPS handshake has begun I should be able to return that. But i cant see from the documentation how you would do that.

**Comment by mzimmerman** [comment]
No, you cannot do that.

Your browser is expecting the https protocol.  You can't just talk http.

Instead, because of security rules of certificates and how TLS works, you have to act like the destination that they're trying to go to.  You have to do that from a CA that is trusted by the browser.  Then and only then will the browser accept your TLS handshake and let you then deliver your HTML through https.

See [whitelistproxy] (https://github.com/mzimmerman/whitelistproxy) for an example of how to do this using goproxy.

**Comment by abourget** [comment]
@mzimmerman 's answer is correct. You can do that with my fork too github.com/abourget/goproxy .. until it's merged back here. @mzimmerman I'm curious, is whitelistproxy a full alternative to goproxy? does it use it internaly ? and if so, what does it add ?

**Comment by mzimmerman** [comment]
Right, it uses goproxy internally.  It manages the whitelist and adds a
user interface for adding new ones.

https://github.com/mzimmerman/whitelistproxy/blob/master/README.md

**Comment by ErikPelli** [comment]
Goproxy HTTPS MITM is now officially supported and part of the master branch. Closing this issue.
